src/Kriti/Eval.hs

```diff
@@ -58,12 +58,14 @@ getSourcePos (RangeError _ pos) = locate pos
 evalPath :: Span -> J.Value -> V.Vector (Accessor) -> ExceptT EvalError (Reader Ctxt) J.Value
 evalPath sp ctx path = do
   src <- asks fst
-  let step :: Monad m => J.Value -> Accessor -> ExceptT EvalError m J.Value
-      step (J.Object o) (Obj _ k _) = maybe (throwError $ InvalidPath src sp path) pure $ Compat.lookup k o
-      step (J.Array xs) (Arr _ i) = maybe (throwError $ InvalidPath src sp path) pure $ xs V.!? i
-      step _ (Obj _ _ _) = throwError $ TypeError src sp "Expected object"
-      step _ (Arr _ _) = throwError $ TypeError src sp "Expected array"
-   in foldlM step ctx path
+  let step :: J.Value -> Accessor -> ExceptT (Either EvalError ()) (Reader Ctxt) J.Value
+      step (J.Object o) (Obj _ NotOptional k _) = maybe (throwError $ Left $ InvalidPath src sp path) pure $ Compat.lookup k o
+      step (J.Object o) (Obj _ Optional k _) = maybe (throwError $ Right ()) pure $ Compat.lookup k o
+      step (J.Array xs) (Arr _ NotOptional i) = maybe (throwError $ Left $ InvalidPath src sp path) pure $ xs V.!? i
+      step (J.Array xs) (Arr _ Optional i) = maybe (throwError $ Right ()) pure $ xs V.!? i
```

Comment:
Wondering if there could be a combinator to help handle the optionality rather than having to have explicit pattern matches insitu. NBD if you want to think about that later.